tests/testLoginManager.php

```diff
@@ -133,4 +173,133 @@ function ( $auth_url, $params ) {
+		}
+
+		// Redirect will have a dynamic cache breaker so cannot assertEquals here.
+		$this->assertNotFalse( strpos( $caught_redirect['location'], 'http://example.org?' ) );
+		$this->assertEquals( 302, $caught_redirect['status'] );
+
+		// Set a one-time login redirect URL.
+		$_REQUEST['redirect_to'] = 'http://example.org/custom';
+
+		// Use the default login redirection.
+		$caught_redirect = [
+			'location' => null,
+			'status'   => null,
+		];
+		try {
+			$login_manager->login_auto();
+		} catch ( Exception $e ) {
+			$caught_redirect = unserialize( $e->getMessage() );
+		}
+
+		// Redirect will have a dynamic cache breaker so cannot assertEquals here.
+		$this->assertNotFalse( strpos( $caught_redirect['location'], 'http://example.org/custom?' ) );
+		$this->assertEquals( 302, $caught_redirect['status'] );
+	}
+
+	/**
+	 * Test that the ULP redirect happens.
+	 */
+	public function testUlpRedirect() {
+		$this->startRedirectHalting();
+
+		$login_manager = new WP_Auth0_LoginManager( self::$users_repo, self::$opts );
+		$this->assertFalse( $login_manager->login_auto() );
+
+		// Activate settings that result in a ULP redirect.
+		self::$opts->set( 'auto_login', 1 );
+		self::auth0Ready( true );
+		self::$opts->set( 'domain', 'test-wp.auth0.com' );
+		$this->assertTrue( WP_Auth0::ready() );
+
+		$caught_redirect = [
+			'location' => null,
+			'status'   => null,
+		];
+		try {
+			// Need to hide error messages here because a cookie is set.
+			// phpcs:ignore
+			@$login_manager->login_auto();
+		} catch ( Exception $e ) {
+			$caught_redirect = unserialize( $e->getMessage() );
+		}
+
+		$this->assertNotFalse( strpos( $caught_redirect['location'], 'https://test-wp.auth0.com/authorize?' ) );
+		$this->assertEquals( 302, $caught_redirect['status'] );
+	}
+
+	/**
+	 * Test that the ULP redirect does not happen under certain conditions.
+	 */
+	public function testThatUlpRedirectDoesNotOccur() {
+		$this->startRedirectHalting();
+
+		$login_manager = new WP_Auth0_LoginManager( self::$users_repo, self::$opts );
+		$this->assertFalse( $login_manager->login_auto() );
+
+		// First, check that a redirect is happening.
+		self::$opts->set( 'auto_login', 1 );
+		self::auth0Ready( true );
+		$this->assertTrue( WP_Auth0::ready() );
+
+		$caught_redirect = [
+			'location' => null,
+			'status'   => null,
+		];
+		try {
+			// Need to hide error messages here because a cookie is set.
+			// phpcs:ignore
+			@$login_manager->login_auto();
+		} catch ( Exception $e ) {
+			$caught_redirect = unserialize( $e->getMessage() );
+		}
+		$this->assertEquals( 302, $caught_redirect['status'] );
+
+		// Test that request method will stop redirect.
+		$req_method_backup         = $_SERVER['REQUEST_METHOD'];
+		$_SERVER['REQUEST_METHOD'] = 'POST';
+		$this->assertFalse( $login_manager->login_auto() );
+		$_SERVER['REQUEST_METHOD'] = $req_method_backup;
+
+		// Test that WP login override will skip the redirect.
```

Comment:
All these 3 checks below should be moved into 3 different tests as the condition for the test changes.